File: tree/Day28/gaoyuyue.java
import java.util.LinkedList;
import java.util.Queue;

class TreeLinkNode {
    int val;
    TreeLinkNode left, right, next;
    TreeLinkNode(int x) { val = x; }
}

public class Solution {
    public void connect(TreeLinkNode root) {
        Queue<TreeLinkNode> firstQueue = new LinkedList<>();
        Queue<TreeLinkNode> secondQueue = new LinkedList<>();
        firstQueue.offer(root);
        TreeLinkNode t;
        while (!firstQueue.isEmpty() || !secondQueue.isEmpty()) {
            if (!firstQueue.isEmpty()) {
                while (!firstQueue.isEmpty()) {
                    t = firstQueue.poll();
                    if (t != null) {
                        if (!firstQueue.isEmpty()){
                            t.next = firstQueue.peek();
                        }
                        secondQueue.offer(t.left);
                        secondQueue.offer(t.right);
                    }
                }
            } else {
                while (!secondQueue.isEmpty()) {
                    t = secondQueue.poll();
                    if (t != null) {
                        if (!secondQueue.isEmpty()){
                            t.next = secondQueue.peek();
                        }
                        firstQueue.offer(t.left);
                        firstQueue.offer(t.right);
                    }
                }
            }
        }
    }
}

File: tree/Day31/gaoyuyue.go
package main

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func addOneRow(root *TreeNode, v int, d int) *TreeNode {
	if d == 1 {
		return &TreeNode{v, root, nil}
	}
	add(root, v, d, 1)
	return root
}

func add(tn *TreeNode, v, d, i int)  {
	if tn == nil {
		return
	}
	if i == d - 1 {
		left := &TreeNode{v, tn.Left, nil}
		right := &TreeNode{v, nil, tn.Right}
		tn.Left = left
		tn.Right = right
		return
	}
	add(tn.Left, v, d, i+1)
	add(tn.Right, v, d, i+1)
}

func main() {
	
}
